### PR TITLE
Update Disassemblers to use Loot Tables, and Use In-Machine interface

### LINKED ISSUE #821: Disassemblers destroys multiple dropped items
**Describe the bug**
Disassemblers triggers for multiple item entities, but only outputs one, deleting the other entities

**To Reproduce**
1. Place a Disassembler
2. Drop many dissassemble-able items on top
3. Lose items :(

**Expected behavior**
Only delete the one item it disassembles

**Fix**
Merge #811 

**Screenshots**

**Environment**
 - Minecraft version: 1.19.2
 - Server version: N/A (private)
 - Installed Datapacks: All/most

### PR DESCRIPTION
**Pull request by jpeterik12**
This PR modernizes Disassemblers, to better fit in with the rest of the machines post machine blocks update. 
It changes Disassemblers to accept input from inside the dropper, rather than item entities above.
It also changes the checking system to one auto generated loot table, rather than many `if data` commands.
It also adds leather armor and a few other damageable tools.

Fixes #821.

### CODE REVIEW

**Comment by BluePsychoRanger** [comment]
Instead of using an id length check, it may be more efficient to put the item in the armor stand's hand and use a loot table to check the item and run the other loot tables. This would be able to utilize the `equipment` predicate and can be put in an `alternatives` condition so it stops when it matches the item id.

File: gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import json
+import urllib.request
+import itertools
+import beet
```
Comment (misode) [inline gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py:4]: Unused import

File: gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    for row in recipe["pattern"]:
+      if len(row) < 2: row = " " + row
+      if len(row) < 3: row = row + " "
+      pattern = pattern + row
+    while (len(pattern) < 9):
+      pattern = pattern + " "
+    pattern = ["".join(g) for k, g in itertools.groupby(pattern)]
+    # for j, patternx in enumerate(pattern):
+    #   if patternx != " " and  len(patternx) != 1:
+    #     pattern[j] = [patternx[0]]*len(patternx)
```
Comment (misode) [inline gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py:42]: Remove commented code if not necessary. Also some more commented code below.

File: gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+leather_items = ["minecraft:leather_helmet","minecraft:leather_chestplate","minecraft:leather_leggings","minecraft:leather_boots"]
+leather_durability = [56, 81, 76, 66]
+other_items = ["minecraft:turtle_helmet","minecraft:bow","minecraft:crossbow","minecraft:shears","minecraft:fishing_rod","minecraft:flint_and_steel"]
+other_durability = [276, 385, 466, 239, 65, 65]
+normal_items = diamond_items + iron_items + gold_items + leather_items + other_items
+normal_durability = diamond_durability + iron_durability + gold_durability + leather_durability + other_durability
+special_items = ["minecraft:shield"]
+special_durability = [337]
+
+recipes_url = "https://raw.githubusercontent.com/InventivetalentDev/minecraft-assets/1.19.2/data/minecraft/recipes/_all.json"
```
Comment (misode) [inline gm4_disassemblers/data/gm4_disassemblers/generate_disassembly.py:23]: Any way to not hardcode the current version would be nice. Also I have a small preference towards <https://raw.githubusercontent.com/misode/mcmeta/1.19.2-summary/data/recipe/data.min.json> :)

File: gm4_disassemblers/data/gm4_disassemblers/functions/main.mcfunction [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # NOTE remove during 1.20 update: updates old machines to include a marker entity
 execute as @e[type=armor_stand,tag=gm4_disassembler] at @s run function gm4_disassemblers:upgrade_machine_stand
 # process machine
-execute as @e[type=marker,tag=gm4_disassembler] at @s positioned ~-0.5 ~0.5 ~-0.5 as @e[type=item,tag=!gm4_disassemble,dx=0,nbt={OnGround:1b}] positioned ~0.5 ~-0.5 ~0.5 if block ~ ~ ~ dropper[triggered=false]{Items:[]} run function #gm4_disassemblers:check_recipe
+execute as @e[type=marker,tag=gm4_disassembler] at @s if block ~ ~ ~ dropper[triggered=false] run function gm4_disassemblers:process
```
[the remark this answers is not in the document]
Comment (jpeterik12) [reply gm4_disassemblers/data/gm4_disassemblers/functions/main.mcfunction:4]: I need to check that there is only one item. I figured if I was going to do two block nbt checks (to check for only one entry in Items), this would be better. If it's not I will adjust.

**Comment by Bloo-dev** [comment]
According to @jpeterik12 this fixes #821.

File: gm4_disassemblers/data/gm4_disassemblers/functions/check_recipe.mcfunction [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# processes each item
+# @s = armor stand display (we need the hand) [tag=gm4_disassembler_stand]
+# located at the disassembler block
+# run from extract
+data modify entity @s HandItems[0] set from storage gm4_disassemblers:temp Items[0]
+execute store result score $damage gm4_disassembler run data get storage gm4_disassemblers:temp Items[0].tag.Damage
+scoreboard players set dropped gm4_disassembler 0
+function #gm4_disassemblers:before_base
+execute if score dropped gm4_disassembler matches 0 run function #gm4_disassemblers:during_base
+execute if score dropped gm4_disassembler matches 0 run function #gm4_disassemblers:after_base
```
Comment (Bloo-dev) [inline gm4_disassemblers/data/gm4_disassemblers/functions/check_recipe.mcfunction:10]: Why did you rename this file to the singular form? This isn't in line with the usual naming convention for custom-crafting type modules.

Comment (jpeterik12) [reply gm4_disassemblers/data/gm4_disassemblers/functions/check_recipe.mcfunction:10]: Honestly, I deleted all the files and just created new ones from scratch. Really just a typo.

**Review by Bloo-dev — APPROVED** [review]
Tested in-game and code reviewed.
